test/common/router/router_test.cc

```diff
@@ -6329,6 +6329,40 @@ TEST_F(RouterTest, ConnectExplicitTcpUpstream) {
   router_.onDestroy();
 }
 
+TEST_F(RouterTest, PostExplicitTcpUpstream) {
+  // Explicitly configure an TCP upstream, to test factory creation.
+  cm_.thread_local_cluster_.cluster_.info_->upstream_config_ =
+      absl::make_optional<envoy::config::core::v3::TypedExtensionConfig>();
+  envoy::extensions::upstreams::http::tcp::v3::TcpConnectionPoolProto tcp_config;
```

Comment:
Ah, you should also be using the generic config here.

envoy::extensions::upstreams::http::tcp::v3::TcpConnectionPoolProto tcp_config;
->
envoy::extensions::upstreams::http::generic::v3::GenericConnectionPoolProto generic_config;

Reply:
thanks, it works!